File: solutions/nim/dnd-character/1/dnd_character.nim
import math, random

randomize()

type
    Character* = object
        strength*: int
        dexterity*: int
        constitution*: int
        intelligence*: int
        wisdom*: int
        charisma*: int
        hitpoints*: int

proc ability*: int =
    var rolls = [0, 0, 0, 0]
    for roll in rolls.mitems:
        roll = rand(1..6)
    sum(rolls) - min(rolls)

proc modifier*(n: int): int =
    int(floor(float(n - 10) / 2))

proc initCharacter*: Character =
    var character: Character
    character.strength = ability()
    character.dexterity = ability()
    character.constitution = ability()
    character.intelligence = ability()
    character.wisdom = ability()
    character.charisma = ability()
    character.hitpoints = 10 + modifier(character.constitution)
    character

File: solutions/lua/dnd-character/1/dnd-character.lua
local Character = {}

local function ability(scores)
    return math.random(3, 18)
end

local function modifier(input)
    return math.floor((input - 10) / 2)
end

function Character:new(name)
    local character = {name = name}
    character.strength = ability()
    character.dexterity = ability()
    character.constitution = ability()
    character.intelligence = ability()
    character.wisdom = ability()
    character.charisma = ability()
    character.hitpoints = 10 + modifier(character.constitution)
    return setmetatable(character, {__index = Character})
end

return {Character = Character, ability = ability, modifier = modifier}
